src/tools.rs

```diff
@@ -267,3 +267,180 @@ pub fn decode_swdord(data: &[u8]) -> i64 {
+    data: &'a [u8],
+}
+
+impl<'a> RawBufReader<'a> {
+    pub fn new(data: &[u8]) -> RawBufReader {
+        RawBufReader { off: 0, data }
+    }
+
+    #[inline]
+    fn ensure(&self, len: usize) -> Option<()> {
+        if self.data.len() < (self.off + len) {
+            return None;
+        }
+        Some(())
+    }
+
+    #[inline]
+    pub fn pos(&self) -> usize {
+        self.off
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.off >= self.data.len()
+    }
+
+    #[inline]
+    pub fn take(&mut self) -> u8 {
+        let b = self.data[self.off];
+        self.off += 1;
+        b
+    }
+
+    #[inline]
+    pub fn peek(&mut self) -> u8 {
+        self.data[self.off]
+    }
+
+    #[inline]
+    pub fn take_slice(&mut self, len: usize) -> Option<&[u8]> {
+        self.ensure(len)?;
+        let s = &self.data[self.off..(self.off + len)];
+        self.off += len;
+        Some(s)
+    }
+
+    #[inline]
+    pub fn decode_leb128_128(&mut self) -> Option<u128> {
+        let mut sz = 0;
+        let mut v: u128 = 0;
+        for c in &self.data[self.off..] {
+            v |= ((c & 0x7f) as u128) << sz;
+            sz += 7;
+            if (c & 0x80) == 0 {
+                self.off += sz / 7;
+                return Some(v);
+            }
+        }
+        None
+    }
+
+    #[inline]
+    pub fn decode_leb128(&mut self) -> Option<u64> {
+        self.decode_leb128_128().map(|v| v as u64)
+    }
+
+    pub fn decode_leb128_128_s(&mut self) -> Option<i128> {
+        let saved_off = self.off;
+        if let Some(v) = self.decode_leb128_128() {
+            let s = self.off - saved_off;
+            let s_mask: u128 = 1 << (s * 7 - 1);
+            return if (v & s_mask) != 0 {
+                // negative
+                Some((v as i128) - (s_mask << 1) as i128)
+            } else {
+                Some(v as i128)
+            };
+        }
+        None
+    }
+
+    pub fn decode_leb128_s(&mut self) -> Option<i64> {
+        self.decode_leb128_128_s().map(|v| v as i64)
+    }
+
+    #[inline(always)]
+    pub fn decode_uhalf(&mut self) -> Option<u16> {
+        self.ensure(2)?;
+        let data = &self.data[self.off..];
+        self.off += 2;
+        Some((data[0] as u16) | ((data[1] as u16) << 8))
+    }
+
+    #[cfg(test)]
+    #[inline(always)]
+    pub fn decode_shalf(&mut self) -> Option<i16> {
+        let uh = self.decode_uhalf()?;
+        Some(if uh >= 0x8000 {
+            ((uh as i32) - 0x10000) as i16
+        } else {
+            uh as i16
+        })
+    }
+
+    #[inline(always)]
+    pub fn decode_uword(&mut self) -> Option<u32> {
+        self.ensure(4)?;
+        let data = &self.data[self.off..];
+        self.off += 4;
+        Some(
+            (data[0] as u32)
+                | ((data[1] as u32) << 8)
+                | ((data[2] as u32) << 16)
+                | ((data[3] as u32) << 24),
```

Comment:
As discussed in person, can we use library functions such [`from_le_bytes`](https://doc.rust-lang.org/std/primitive.u32.html#method.from_le_bytes) instead?

Reply:
Sure!